usb/esp_tinyusb/tusb_msc_storage_sdmmc.c

```diff
@@ -0,0 +1,276 @@
+    }
+    return ESP_OK;
+
+fail:
+    if (fs) {
+        f_mount(NULL, drv, 0);
+    }
+    esp_vfs_fat_unregister_path(base_path);
+    ff_diskio_unregister(pdrv);
+    return err;
+}
+
+esp_err_t tinyusb_msc_storage_mount(const char *base_path)
+{
+    if (s_fat_mounted) {
+        return ESP_OK;
+    }
+
+    ESP_LOGI(TAG, "Initializing FAT");
+
+    // Options for mounting the filesystem.
+    // If format_if_mount_failed is set to true, SD card will be partitioned and
+    // formatted in case when mounting fails.
+    esp_vfs_fat_sdmmc_mount_config_t mount_config = {
+        .format_if_mount_failed = true,
+        .max_files = 5,
+        .allocation_unit_size = 16 * 1024
+    };
+
+    // connect driver to FATFS
+    BYTE pdrv = 0xFF;
+    ESP_RETURN_ON_ERROR(ff_diskio_get_drive(&pdrv), TAG,
+                        "The maximum count of volumes is already mounted");
+
+    ESP_LOGI(TAG, "using pdrv=%i", pdrv);
+
+    ESP_RETURN_ON_ERROR(mount_to_vfs_fat(&mount_config, s_card, pdrv, base_path), TAG,
+                        "Failed to mount storage");
+
+    s_fat_mounted = true;
+    s_base_path = base_path;
+    return ESP_OK;
+}
+
+static esp_err_t unmount_card_core(const char *base_path, sdmmc_card_t *card)
+{
+    BYTE pdrv = ff_diskio_get_pdrv_card(card);
+    if (pdrv == 0xff) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    // unmount
+    char drv[3] = {(char)('0' + pdrv), ':', 0};
+    f_mount(0, drv, 0);
+    // release SD driver
+    ff_diskio_unregister(pdrv);
+
+    return esp_vfs_fat_unregister_path(base_path);
+}
+
+esp_err_t tinyusb_msc_storage_unmount(void)
+{
+    if (!s_fat_mounted) {
+        return ESP_OK;
+    }
+
+    esp_err_t err = unmount_card_core(s_base_path, s_card);
+
+    s_base_path = NULL;
+    s_fat_mounted = false;
+
+    return err;
+
+}
+
+uint32_t tinyusb_msc_storage_get_sector_count(void)
+{
+    assert(s_card);
+
+    return (uint32_t)s_card->csd.capacity;
+}
+
+uint32_t tinyusb_msc_storage_get_sector_size(void)
+{
+    assert(s_card);
+
+    return (uint32_t)s_card->csd.sector_size;
+}
+
+esp_err_t tinyusb_msc_storage_read_sector(uint32_t lba, uint32_t offset, size_t size, void *dest)
+{
+    assert(s_card);
+    return sdmmc_read_sectors(s_card, dest, lba, size / tinyusb_msc_storage_get_sector_size());
+}
+
+esp_err_t tinyusb_msc_storage_write_sector(uint32_t lba, uint32_t offset, size_t size, const void *src)
+{
+    assert(s_card);
+
+    if (s_fat_mounted) {
+        ESP_LOGE(TAG, "can't write, FAT mounted");
+        return ESP_ERR_INVALID_STATE;
+    }
+    size_t addr = lba * tinyusb_msc_storage_get_sector_size() + offset; // Address of the data to be read, relative to the beginning of the partition.
+    size_t sector_size = s_card->csd.sector_size;
+    if (addr % sector_size != 0 || size % sector_size != 0) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    ESP_RETURN_ON_ERROR(sdmmc_erase_sectors(s_card, lba, size / tinyusb_msc_storage_get_sector_size(), SDMMC_ERASE_ARG),
```

Comment:
Unlike with SPI flash, it's not necessary to erase SD card sectors before writing — this is done by the card itself if necessary.

Reply:
ok, I have removed the erase API from sd card code.